pkg/controller/beat/common/pod.go

```diff
@@ -0,0 +1,143 @@
+		Limits: map[corev1.ResourceName]resource.Quantity{
+			corev1.ResourceMemory: resource.MustParse("200Mi"),
+			corev1.ResourceCPU:    resource.MustParse("100m"),
+		},
+		Requests: map[corev1.ResourceName]resource.Quantity{
+			corev1.ResourceMemory: resource.MustParse("200Mi"),
+			corev1.ResourceCPU:    resource.MustParse("100m"),
+		},
+	}
+)
+
+func buildPodTemplate(
+	params DriverParams,
+	defaultImage container.Image,
+	modifyPodFunc func(builder *defaults.PodTemplateBuilder),
+	configHash hash.Hash) corev1.PodTemplateSpec {
+	podTemplate := params.GetPodTemplate()
+
+	// Token mounting gets defaulted to false, which prevents from detecting whether user had set it.
+	// Instead, checking that here, before the default is applied.
+	// This is required for autodiscover which is enabled by default.
+	if podTemplate.Spec.AutomountServiceAccountToken == nil {
+		t := true
+		podTemplate.Spec.AutomountServiceAccountToken = &t
+	}
+
+	spec := &params.Beat.Spec
+	builder := defaults.NewPodTemplateBuilder(podTemplate, spec.Type)
+
+	// might be nil if caller wants to use the default builder without any modifications
+	if modifyPodFunc != nil {
+		modifyPodFunc(builder)
+	}
+
+	builder = builder.
+		WithTerminationGracePeriod(30).
+		WithEnv(corev1.EnvVar{
+			Name: "NODE_NAME",
+			ValueFrom: &corev1.EnvVarSource{
+				FieldRef: &corev1.ObjectFieldSelector{
+					FieldPath: "spec.nodeName",
+				},
+			}}).
+		WithResources(defaultResources).
+		WithHostNetwork().
+		WithLabels(map[string]string{
+			ConfigChecksumLabel: fmt.Sprintf("%x", configHash.Sum(nil)),
+			VersionLabelName:    spec.Version}).
+		WithDockerImage(spec.Image, container.ImageRepository(defaultImage, spec.Version)).
+		WithArgs("-e", "-c", ConfigMountPath).
+		WithDNSPolicy(corev1.DNSClusterFirstWithHostNet).
+		WithPodSecurityContext(corev1.PodSecurityContext{
+			RunAsUser: pointer.Int64(0),
+		})
+
+	if ShouldManageAutodiscoverRBAC() {
+		autodiscoverServiceAccountName := ServiceAccountName(params.Beat.Name)
+		// If SA is already provided, the call will be no-op. This is fine as we then assume
+		// that for this resource (despite operator configuration) the user took the responsibility
+		// of configuring RBAC.
+		builder.WithServiceAccount(autodiscoverServiceAccountName)
+	}
+
+	dataVolume := createDataVolume(params)
+	volumes := []volume.VolumeLike{
+		volume.NewSecretVolume(
+			ConfigSecretName(spec.Type, params.Beat.Name),
+			ConfigVolumeName,
+			ConfigMountPath,
+			ConfigFileName,
+			0600),
+		dataVolume,
+	}
+
+	if params.Beat.AssociationConf().CAIsConfigured() {
+		volumes = append(volumes, volume.NewSecretVolumeWithMountPath(
+			params.Beat.AssociationConf().GetCASecretName(),
+			CAVolumeName,
+			CAMountPath))
+	}
+
+	for _, v := range volumes {
+		builder = builder.WithVolumes(v.Volume()).WithVolumeMounts(v.VolumeMount())
+	}
+
+	builder = builder.WithLabels(commonhash.SetTemplateHashLabel(NewLabels(params.Beat), builder.PodTemplate))
```

Comment:
Why do we need the hash label at the pod level? If the pod spec changes the Deployment/DaemonSet controller handles, if the config changes we have the configchecksum label set.
I understand why we'd have it at the DaemonSet or Deployment level though, so we can easily compare and reconcile.